src/Word/CamelCaseToSeparator.php

```diff
@@ -4,43 +4,42 @@
 
 namespace Laminas\Filter\Word;
 
-use Closure;
-use Laminas\Stdlib\StringUtils;
+use Laminas\Filter\FilterInterface;
+use Laminas\Filter\ScalarOrArrayFilterCallback;
 
 use function preg_replace;
 
 /**
  * @psalm-type Options = array{
  *     separator?: string,
- *     ...
  * }
  * @template TOptions of Options
- * @extends AbstractSeparator<TOptions>
+ * @implements FilterInterface<string|array<array-key, string|mixed>>
  */
-class CamelCaseToSeparator extends AbstractSeparator
+final class CamelCaseToSeparator implements FilterInterface
 {
-    public function filter(mixed $value): mixed
+    private readonly string $separator;
+
+    /** @param Options $options */
+    public function __construct(array $options = [])
     {
-        return self::applyFilterOnlyToStringableValuesAndStringableArrayValues(
-            $value,
-            Closure::fromCallable([$this, 'filterNormalizedValue'])
-        );
+        $this->separator = $options['separator'] ?? ' ';
+    }
+
+    public function __invoke(mixed $value): mixed
+    {
+        return $this->filter($value);
     }
 
-    /**
-     * @param  string|string[] $value
-     * @return string|string[]
-     */
-    private function filterNormalizedValue(string|array $value): string|array
+    public function filter(mixed $value): mixed
     {
-        if (StringUtils::hasPcreUnicodeSupport()) {
-            $pattern     = ['#(?<=(?:\p{Lu}))(\p{Lu}\p{Ll})#', '#(?<=(?:\p{Ll}|\p{Nd}))(\p{Lu})#'];
-            $replacement = [$this->separator . '\1', $this->separator . '\1'];
-        } else {
-            $pattern     = ['#(?<=(?:[A-Z]))([A-Z]+)([A-Z][a-z])#', '#(?<=(?:[a-z0-9]))([A-Z])#'];
-            $replacement = ['\1' . $this->separator . '\2', $this->separator . '\1'];
-        }
-
-        return preg_replace($pattern, $replacement, $value);
+        return ScalarOrArrayFilterCallback::applyRecursively(
+            $value,
+            fn (string $input): string => preg_replace(
+                ['#(?<=(?:\p{Lu}))(\p{Lu}\p{Ll})#', '#(?<=(?:\p{Ll}|\p{Nd}))(\p{Lu})#'],
+                [$this->separator . '\1', $this->separator . '\1'],
+                $input
+            )
+        );
```

Comment:
```suggestion
    $pattern = <<<REGEXP
        /
        (
            (?:\p{Lu}\p{Ll}+) # Upper followed by lower
            |
            (?:\p{Lu}+(?!\p{Ll})) # Upper not followed by lower
            |
            (?:\p{N}+) # Runs of numbers
        )
        /ux
        REGEXP;

        return ScalarOrArrayFilterCallback::applyRecursively(
            $value,
            fn (string $input): string => implode(
                $this->separator,
                preg_split(
                    $pattern,
                    str_replace($this->separator, '', $input),
                    -1,
                    PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY,
                ),
            ),
        );
```

Something like this should work OK, obviously we need the extra test cases - not sure how to handle "invalid" string input here - i.e. where the separator or other punctuation is present. IMO, strange output for inputs such as `Foo-Bar!_Ding_Dong` is the end users fault, so perhaps drop the `str_replace` in favour of `Foo-Bar => Foo---Bar`

Reply:
That'd make sense. People still have the option of pre-stripping the seperator if they have a need, so might be better than forcing it

Reply:
I still need to look at the migration docs, and I'm unlikely to get back to it today, but this is where I'm at